FlowCrypt/Controllers/Compose/ComposeViewController.swift

```diff
@@ -328,104 +325,112 @@ extension ComposeViewController {
+        //  let's find all keys that match and save the pass phrase for all
+        let allKeys = try await self.keyService.getPrvKeyInfo()
+        guard allKeys.isNotEmpty else {
+            // tom - todo - nonsensical error type choice https://github.com/FlowCrypt/flowcrypt-ios/issues/859
+            //   I copied it from another usage, but has to be changed
+            throw KeyServiceError.retrieve
+        }
+        let matchingKeys = try await self.keyMethods.filterByPassPhraseMatch(keys: allKeys, passPhrase: passPhrase)
+        // save passphrase for all matching keys
+        self.passPhraseService.savePassPhrasesInMemory(passPhrase, for: matchingKeys)
+        // now figure out if the pass phrase also matched the signing prv itself
+        let matched = matchingKeys.first(where: { $0.fingerprints.first == signingKey.fingerprints.first })
+        return matched != nil// true if the pass phrase matched signing key
     }
 
-    private func sendMessage(_ signingKey: PrvKeyInfo) {
+    private func sendMessage(_ signingKey: PrvKeyInfo) async throws {
         view.endEditing(true)
         navigationItem.rightBarButtonItem?.isEnabled = false
 
         let spinnerTitle = contextToSend.attachments.isEmpty ? "sending_title" : "encrypting_title"
         showSpinner(spinnerTitle.localized)
 
         let selectedRecipients = contextToSend.recipients.filter(\.state.isSelected)
-        selectedRecipients.forEach(evaluate)
+        for selectedRecipient in selectedRecipients {
+            evaluate(recipient: selectedRecipient)
+        }
 
         // TODO: - fix for spinner
         // https://github.com/FlowCrypt/flowcrypt-ios/issues/291
-        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) { [weak self] in
-            guard let self = self else { return }
-            let result = self.composeMessageService.validateMessage(
-                input: self.input,
-                contextToSend: self.contextToSend,
-                email: self.email,
-                signingPrv: signingKey
-            )
-            switch result {
-            case .success(let message):
-                self.encryptAndSend(message)
-            case .failure(let error):
-                self.handle(error: error)
-            }
-        }
-    }
-
-    private func encryptAndSend(_ message: SendableMsg) {
-        Task {
-            do {
-                try await service.encryptAndSend(message: message,
-                                                 threadId: input.threadId,
-                                                 progressHandler: { [weak self] progress in
-                    self?.updateSpinner(progress: progress)
-                })
-                handleSuccessfullySentMessage()
-            } catch {
-                if let error = error as? ComposeMessageError {
-                    handle(error: error)
-                }
+        try await Task.sleep(nanoseconds: 100 * 1_000_000) // 100ms
```

Comment:
@tomholub 
Why we need such delays?

Reply:
Earlier there was some problem with spinner rendering. Maybe the spinner didn't get rendered before the app got "busy" and then it would remain not rendered until the work was done. Unsure.